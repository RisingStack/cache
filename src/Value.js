/* @flow */

export default class Value<V> {
  value: V;
  options: TTLOptions;
  createdAt: Date;
  expireAt: Date;
  staleAt: Date;

  constructor(value: V, options: TTLOptions = {}) {
    this.value = value
    this.createdAt = new Date()
    this.refresh(options)
  }

  refresh(options: TTLOptions = {}) {
    this.options = options
    const { expire, stale } = options
    const now = Date.now()

    if (Number.isInteger(expire)) {
      this.expireAt = new Date(now + expire)
    }

    if (Number.isInteger(stale)) {
      this.staleAt = new Date(now + stale)
    }
  }

  isExpired(): boolean {
    if (typeof this.expireAt === 'undefined') {
      return false
    }

    return this.expireAt.valueOf() < Date.now()
  }

  isStale(): boolean {
    if (typeof this.staleAt === 'undefined') {
      return false
    }

    return this.staleAt.valueOf() < Date.now()
  }

  toJSON(): string {
    return JSON.stringify({
      value: this.value,
      options: this.options
    })
  }

  static fromJSON(json: string): ?Value<V> {
    try {
      const { value, options } = JSON.parse(json)
      return new Value(value, options)
    } catch (err) {
      return null
    }
  }
}
