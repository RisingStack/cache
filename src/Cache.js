/* @flow */

import Value from './Value'

export default class Cache<K, V> {
  stores: Array<Store<K, V>>

  constructor(stores: Array<Store<K, V>>) {
    if (!Array.isArray(stores) || !stores.length) {
      throw new TypeError('`stores` must be an Array with at least one Store')
    }

    this.stores = stores
  }

  async get(key: K): Promise<?Value<V>> {
    // eslint-disable-next-line no-restricted-syntax
    for (const store of this.stores) {
      // eslint-disable-next-line no-await-in-loop
      const value = await store.get(key)
      if (value) {
        return value
      }
    }

    return undefined
  }

  async set(key: K, value: V, options: TTLOptions): Promise<Array<*>> {
    return Promise.all(this.stores.map((store) => store.set(key, value, options)))
  }

  clear() {
    this.stores.forEach((store) => store.clear())
  }

  async refresh(key: K, func: (K | void) => Promise<V> | V, options: TTLOptions): Promise<V> {
    const value = await func(key)

    process.nextTick(() => {
      this.stores.forEach((store) => store.set(key, value, options))
    })

    return value
  }

  async wrap(key: K, func: (K | void) => Promise<V> | V, options: TTLOptions): Promise<?V> {
    const value = await this.get(key)

    // try to refresh if not in cache or the value is expired
    if (!value || value.isExpired()) {
      try {
        const result = await this.refresh(key, func, options)
        return result
      } catch (ex) {
        // throw error if the value is outdated
        if (!value || (value && value.isStale())) {
          throw ex
        }
        // ignore if stale
        return value.value
      }
    }

    return value.value
  }

  getStats(): StoreStats {
    return this.stores.reduce((stats, store) => {
      const { getCount, hitCount } = store.getStats()
      return {
        getCount: stats.getCount + getCount,
        hitCount: stats.hitCount + hitCount
      }
    }, { getCount: 0, hitCount: 0 })
  }
}
