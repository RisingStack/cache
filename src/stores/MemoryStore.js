/* @flow */

import Value from '../Value'

export default class MemoryStore<V> {
  store: Map<string, Value<V>>;
  stats: StoreStats;

  constructor() {
    this.store = new Map()
    this.stats = {
      getCount: 0,
      hitCount: 0
    }
  }

  get(key: string): Promise<?Value<V>> {
    const value = this.store.get(key)

    this.stats.getCount += 1
    if (value) {
      this.stats.hitCount += 1
    }

    return Promise.resolve(value)
  }

  set(key: string, value: V, options: TTLOptions): Promise<Value<V>> {
    const val = new Value(value, options)
    this.store.set(key, val)
    return Promise.resolve(val)
  }

  delete(key: string): Promise<*> {
    this.store.delete(key)
    return Promise.resolve()
  }

  clear() {
    this.store.clear()
  }

  async size(): Promise<number> {
    return Promise.resolve(this.store.size)
  }

  getStats(): StoreStats {
    return this.stats
  }

  resetStats(): StoreStats {
    this.stats = {
      getCount: 0,
      hitCount: 0
    }

    return this.getStats()
  }
}
