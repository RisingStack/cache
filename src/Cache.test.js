import Cache from './Cache'
import Value from './Value'
import MemoryStore from './stores/MemoryStore'

describe('Cache', () => {
  let stores
  let cache

  beforeEach(() => {
    stores = [new MemoryStore(), new MemoryStore()]
    cache = new Cache(stores)

    jest.useFakeTimers()
  })

  it('should throw an error if constructed with no stores', () => {
    const error = new TypeError('`stores` must be an Array with at least one Store')
    expect(() => new Cache()).toThrowError(error)
    expect(() => new Cache([])).toThrowError(error)
    expect(() => new Cache(new MemoryStore())).toThrowError(error)
  })

  describe('.get()', () => {
    it('should return a Promise<Value>', async () => {
      let value = await cache.get('key')
      expect(value).toBeUndefined()

      const item = { val: 1 }
      cache.set('key', item)

      value = await cache.get('key')
      expect(value).toBeInstanceOf(Value)
      expect(value).toMatchObject({ value: item })

      expect(cache).toMatchSnapshot()
    })

    it('should check for the value in every store', async () => {
      expect(cache).toMatchSnapshot()
      expect(stores.map((store) => store.getStats())).toEqual([{
        getCount: 0,
        hitCount: 0
      }, {
        getCount: 0,
        hitCount: 0
      }])

      await cache.get('key')

      expect(cache).toMatchSnapshot()
      expect(stores.map((store) => store.getStats())).toEqual([{
        getCount: 1,
        hitCount: 0
      }, {
        getCount: 1,
        hitCount: 0
      }])

      stores[1].set('key', { val: 1 }, { expire: 100 })

      await cache.get('key')

      expect(cache).toMatchSnapshot()
      expect(stores.map((store) => store.getStats())).toEqual([{
        getCount: 2,
        hitCount: 0
      }, {
        getCount: 2,
        hitCount: 1
      }])
    })

    it('should emit an error when a store fails to get an item', async () => {
      const error = new Error('internal error')
      stores[0].get = () => Promise.reject(error)
      cache.emit = jest.fn()

      const key = 'key'
      await cache.get(key)

      expect(cache.emit).toHaveBeenCalledWith('error', 'Failed to get an item from store', {
        key,
        error
      })
    })
  })

  describe('.set()', () => {
    it('should set the value in every store', async () => {
      const item = { val: 1 }
      const options = { expire: 10, stale: 5 }
      cache.set('key', item, options)

      const values = await Promise.all(stores.map((store) => store.get('key')))
      expect(values[0]).toBeInstanceOf(Value)
      expect(values[0]).toMatchObject({ options, value: item })
      expect(values[1]).toBeInstanceOf(Value)
      expect(values[1]).toMatchObject({ options, value: item })
    })
  })

  describe('.clear()', () => {
    it('should clear all the stores', async () => {
      cache.set('key', { val: 1 })
      let sizes = await Promise.all(stores.map((store) => store.size()))
      expect(sizes).toEqual([1, 1])

      cache.clear()

      sizes = await Promise.all(stores.map((store) => store.size()))
      expect(sizes).toEqual([0, 0])
    })
  })

  describe('.refresh()', () => {
    it('should refresh the cache with the returned value', async () => {
      const item = { val: 1 }
      const options = { expire: 10, stale: 5 }
      const wrappedFunction = jest.fn(() => Promise.resolve(item))

      const result = await cache.refresh('key', wrappedFunction, options)
      expect(result).toEqual(item)
      jest.runAllTicks()

      const values = await Promise.all(stores.map((store) => store.get('key')))
      expect(values[0]).toMatchObject({ options, value: item })
      expect(values[1]).toMatchObject({ options, value: item })

      expect(cache).toMatchSnapshot()
    })
  })

  describe('.wrap', () => {
    it('should return the value if it is in the cache', async () => {
      const item = { val: 1 }
      const wrappedFunction = jest.fn(() => Promise.resolve(item))

      cache.set('key', item, { expire: 100, stale: 100 })

      const result = await cache.wrap('key', wrappedFunction)
      expect(result).toEqual(item)
      expect(wrappedFunction).not.toHaveBeenCalled()
    })

    it('should refresh the value if it is not in the cache', async () => {
      const item = { val: 1 }
      const wrappedFunction = jest.fn(() => Promise.resolve(item))

      expect(cache).toMatchSnapshot()

      const result = await cache.wrap('key', wrappedFunction)
      expect(result).toEqual(item)
      expect(wrappedFunction).toHaveBeenCalled()

      jest.runAllTicks()

      expect(cache).toMatchSnapshot()
    })

    it('should refresh the value if it is not in the cache (with cacheOptions)', async () => {
      const item = { val: 1 }
      const wrappedFunction = jest.fn(() => Promise.resolve({ value: item, cacheOptions: { expire: 10, stale: 5 } }))

      expect(cache).toMatchSnapshot()

      const result = await cache.wrap('key', wrappedFunction)
      expect(result).toEqual(item)
      expect(wrappedFunction).toHaveBeenCalled()

      jest.runAllTicks()

      expect(cache).toMatchSnapshot()
    })

    it('should refresh the value if it is expired', async () => {
      const item = { val: 1 }
      const wrappedFunction = jest.fn(() => Promise.resolve(item))
      cache.set('key', item)
      const value = await cache.get('key')
      value.isExpired = () => true

      expect(cache).toMatchSnapshot()

      const result = await cache.wrap('key', wrappedFunction)
      expect(result).toEqual(item)
      expect(wrappedFunction).toHaveBeenCalled()

      jest.runAllTicks()

      expect(cache).toMatchSnapshot()
    })

    it('should return the value if it is expired but stale and the wrapped function fails', async () => {
      const item = { val: 1 }
      const wrappedFunction = jest.fn(() => Promise.reject(new Error('error')))
      cache.set('key', item)
      const value = await cache.get('key')
      value.isExpired = () => true
      value.isStale = () => false

      expect(cache).toMatchSnapshot()

      const result = await cache.wrap('key', wrappedFunction)
      expect(result).toEqual(item)
      expect(wrappedFunction).toHaveBeenCalled()

      jest.runAllTicks()

      expect(cache).toMatchSnapshot()
    })

    it('should throw an error if the item is not in cache and the wrapped function fails', async () => {
      const error = new Error('error')
      const wrappedFunction = jest.fn(() => Promise.reject(error))

      let ex
      try {
        await cache.wrap('key', wrappedFunction)
      } catch (err) {
        ex = err
      }

      expect(ex).toMatchSnapshot()
    })

    it('should throw an error if the value in cache is expired and stale and the wrapped function fails', async () => {
      const error = new Error('error')
      const wrappedFunction = jest.fn(() => Promise.reject(error))
      cache.set('key', { val: 1 })
      const value = await cache.get('key')
      value.isExpired = () => true
      value.isStale = () => true

      let ex
      try {
        await cache.wrap('key', wrappedFunction)
      } catch (err) {
        ex = err
      }

      expect(ex).toMatchSnapshot()
    })
  })
})
