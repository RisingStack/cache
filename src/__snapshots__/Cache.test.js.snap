exports[`Cache .get() should check for the value in every store 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [],
    },
  ],
}
`;

exports[`Cache .get() should check for the value in every store 2`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 0,
      },
      "store": Array [],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 0,
      },
      "store": Array [],
    },
  ],
}
`;

exports[`Cache .get() should check for the value in every store 3`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 2,
        "hitCount": 0,
      },
      "store": Array [],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 2,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .get() should return a Promise<Value> 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 2,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .refresh() should refresh the cache with the returned value 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{\"expire\":10,\"stale\":5}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{\"expire\":10,\"stale\":5}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should refresh the value if it is expired 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should refresh the value if it is expired 2`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 2,
        "hitCount": 2,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should refresh the value if it is not in the cache 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [],
    },
  ],
}
`;

exports[`Cache .wrap should refresh the value if it is not in the cache 2`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should return the value if it is expired but stale and the wrapped function fails 1`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 1,
        "hitCount": 1,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should return the value if it is expired but stale and the wrapped function fails 2`] = `
Cache {
  "stores": Array [
    MemoryStore {
      "stats": Object {
        "getCount": 2,
        "hitCount": 2,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
    MemoryStore {
      "stats": Object {
        "getCount": 0,
        "hitCount": 0,
      },
      "store": Array [
        Array [
          "key",
          "{\"value\":{\"val\":1},\"options\":{}}",
        ],
      ],
    },
  ],
}
`;

exports[`Cache .wrap should throw an error if the item is not in cache and the wrapped function fails 1`] = `[Error: error]`;

exports[`Cache .wrap should throw an error if the value in cache is expired and stale and the wrapped function fails 1`] = `[Error: error]`;
